lib/instrumentation/azure-functions.js

```diff
@@ -363,30 +363,55 @@ function instrument(agent) {
       isFirstRun = false;
     }
 
+    // In programming model v3 the InvocationContext includes
+    // `bindingDefinitions` and `executionContext`. In v4 the structure is a
+    // little different.
+    let bindingDefinitions = context.bindingDefinitions;
+    if (!bindingDefinitions) {
+      bindingDefinitions = [];
+      // Input bindings
+      bindingDefinitions.push({
+        name: context?.options?.trigger?.name,
+        type: context?.options?.trigger?.type,
+        direction: context?.options?.trigger?.direction,
+      });
+      // Output bindings
+      bindingDefinitions.push(context?.options?.return);
+    }
+    let executionContext = context.executionContext;
+    if (!executionContext) {
+      executionContext = {
+        functionDirectory: '',
+        functionName: context.functionName,
+      };
+    }
+
     const funcInfo = (hookCtx.hookData.funcInfo = new FunctionInfo(
-      context.bindingDefinitions,
-      context.executionContext,
+      bindingDefinitions,
+      executionContext,
       log,
     ));
     const triggerType = funcInfo.triggerType;
 
-    // Handle trace-context.
-    // Note: We ignore the `context.traceContext`. By default it is W3C
-    // trace-context that continues the given traceparent in headers. However,
-    // we do not injest that span, so would get a broken distributed trace if
-    // we included it.
+    // `InvocationContext.traceContext` is broken: it results in sampled=false
+    // (i.e. traces are discarded) and/or broken traces because it creates an
+    // internal Span ID in the trace that cannot be ingested.
+    // See this for full explanation:
+    // https://github.com/elastic/apm-agent-nodejs/pull/4426#issuecomment-2596922653
     let traceparent;
     let tracestate;
-    if (triggerType === TRIGGER_HTTP && context.req && context.req.headers) {
-      traceparent =
-        context.req.headers.traceparent ||
-        context.req.headers['elastic-apm-traceparent'];
```

Comment:
Reviewer note: I'm also dropping support for the now long-ancient 'elastic-apm-traceparent' header. This is obsolete.